Add get-coord API and general query example (#7964)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 6 specification.  All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

== Status
This is an experimental extension specification, intended to provide early
access to features and gather community feedback.  Interfaces defined in this
specification are implemented in {dpcpp}, but they are not finalized and may
change incompatibly in future versions of {dpcpp} without prior notice.
*Shipping software products should not rely on APIs defined in this
specification.*

== Backend support status
This extension is currently implemented in {dpcpp} only for devices
that contain a matrix hardware, specifically Intel(R) Advanced Matrix
Extensions (Intel(R) AMX), Intel(R) Xe Matrix Extensions (Intel(R)
XMX) and Nvidia(R) Tensor Cores.

> The `joint_matrix_mad` function is an optional kernel feature as defined

Author: Yes, it is accurate. We cannot detect whether the kernel is supported or not just by looking at joint matrix types individually. "joint_matrix_mad" is the one that include the 3 matrices so we can have the full used combination (sizes and types) from there.

> The `joint_matrix_mad` function is an optional kernel feature as defined

Reviewer: It seems like both the `joint_matrix` type and the `joint_matrix_mad` function are "optional kernel features".  Imagine a kernel that creates a `joint_matrix` type and does *not* call `joint_matrix_mad` at all.  Is such a kernel guaranteed to run on all possible devices?  I'm guessing that it will probably not.

I think there should probably be two rules to determine whether a kernel is supported on a device:

* Every `joint_matrix` type declaration must be one of the supported types according to the device's query, and
* Every `joint_matrix_mad` function call must use a combination of types that is supported according to the device's query.

If you agree, I think the wording can be tweaked like:

> The `joint_matrix` type and the `joint_matrix_mad` function are optional kernel features as defined ...

and:

> If the application submits a kernel using an unsupported `joint_matrix` type or calls `joint_matrix_mad` with an unsupported combination, ...

implementation throws a synchronous exception with the
`errc::kernel_not_supported` error code as described in section 5.7.

== Overview
Joint matrix is a SYCL extension for matrix hardware programming. It
unifies targets like Intel AMX in CPUs, Intel XMX in Intel GPUs and
Nvidia Tensor Cores. This provides a portable and performant API for
users who want to build their own neural networks applications,
perform custom optimzations, or experiment with new operations in a
timely and performing manner.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification. An implementation supporting this extension must predefine
the macro `SYCL_EXT_ONEAPI_MATRIX` to one of the values defined in the
table below. Applications can test for the existence of this macro to
determine if the implementation supports this feature, or applications
can test the macro's value to determine which of the extension's
features the implementation supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

> === New `joint_matrix` class

Author: The group argument on the matrix specifies that the matrix is shared among the work items. The description you provided is only about the operations not the data structure itself.
For instance, we can still have group operations (load, store, mad) but they operate on per WI data structure. In this case, the matrix shape will reflect the WI portion.

We introduce a new class called `joint_matrix`. The user needs to
specify the group memory scope, the type of the elements, the shape,
the matrix use, and the memory layout of the matrix. This results in
the following description:

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
          layout Layout = (Use == use::accumulator) ?
          layout::dynamic : /*unspecified*/ >

> struct joint_matrix;

Reviewer: I just noticed that you removed the body of the `joint_matrix` structure from the specification.  I think we need to specify what constructors and assignment operators it supports.  This would be a minimal set, which is probably all we need:

```
template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
          layout Layout = (Use == use::accumulator) ?
          layout::dynamic : /*unspecified*/ >
struct joint_matrix {
  joint_matrix();
  joint_matrix(const joint_matrix &);
  joint_matrix &operator=(const joint_matrix &);
};
```

The spec should also say whether these constructors / copy-assignment-operator are "group functions" that must be encountered in convergent code.  This is somewhat related to `joint_matrix_apply`.  If that function is not a group function, then the constructors / copy-assignment-operator probably should not be either.

Author: We don't support the assignment operator and copy constructor.
You cannot do something like:
joint_matrix<> A;
joint_matrix<> B;
A=B;
If we would enable that, we will have to provide a separate API for this.
In the meantime, use joint_matrix_apply to assign/copy elements of the two matrices.

} // namespace sycl::ext::oneapi::experimental::matrix
```
When the `Use` parameter is `use::accumulator`, the `Layout` parameter
defaults to `layout::dynamic`, and it is invalid to specify any other
value for `Layout`. When `Use` has any other value, there is no default
for `Layout`, and the application must specify one explicitly.

==== Use
The main operation performed by the matrix hardware is `D=C+A*B`. `Use`
argument specifies the usage of the matrix: matrix left (`A`), matrix
right (`B`) or accumulator (`C`) and `D`. This is required by backend
implementations to reason about the layout of the matrix in registers.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

enum class use {
  a,
  b,
  accumulator
};

} // namespace sycl::ext::oneapi::experimental::matrix
```

==== Shape
The shape of a `joint_matrix` refers to its number of rows `Rows` and
number of columns `Cols`.

==== Layout
This specifies the memory layout and it can be row major or column
major. `dynamic` layout is used on the `joint_matrix` type when this
is specified on the memory operations instead for the `accumulator`

> matrix.

Author: I already added that to the "New joint_matrix class" section. I will move it to layout section then.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

enum class layout {
  row_major,
  col_major,
  dynamic
};

} // namespace sycl::ext::oneapi::experimental::matrix
```

==== Group Memory Scope
In this API, we use the terminology of `joint_matrix` instead of plain
`matrix` to emphasize that the matrix is shared among a group of work
items and is not private to each work item. The group scope is added
as an additional template parameter. `Group` template parameter must
be `sycl::sub_group`. In this case, a matrix is declared as follows:

```c++
joint_matrix<sub_group, int8_t, use::a, tM, tN, layout::row_major> tA;
```

> === Matrix Operations and their Execution Scope

Author: Sounds good.
Yes element wise operations (joint_matrix_apply) is a group function. So I kept it as a subsection.

`layout::dynamic`.
The second overload without a memory layout must not be used with a
`joint_matrix` type that has `layout::dynamic`.

The base pointer `src` of type `T` here determines the starting address of the
matrix to be loaded from. `Layout` determines whether the data is
being read in a row (`row_major`), column major (`col_major`)
fashion. `stride` describes the number of elements between consecutive
rows for the row major layout, or between columns for the column major
layout.

==== Store
```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, typename T, size_t Rows, size_t Cols,
          access::address_space Space, access::decorated IsDecorated>
  void joint_matrix_store(Group g,
    joint_matrix<Group, T, use::accumulator, Rows, Cols,
    layout::dynamic> &res,
    multi_ptr<T, Space, IsDecorated> dest, size_t stride, layout Layout);

} // namespace sycl::ext::oneapi::experimental::matrix
```
This function stores the data in the accumulator matrix from the 2d
tiles back to memory.

The base pointer `dest` here determines the starting address of the
matrix to be stored. `Layout` determines whether the data is being

> written in a row (`row_major`), column major (`column_major`)

Reviewer: ```suggestion
written in a row (`row_major`), column major (`col_major`)
```

fashion. `stride` describes the number of elements between consecutive
rows for the row major layout, or between columns for the column major layout.

==== Multiply and Add

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, typename Ta, typename Tb, typename Tc,
  std::size_t M, std::size_t K, std::size_t N,
            layout LayoutA, layout LayoutB>
  joint_matrix<Group, Td, use::accumulator, M, N, layout::dynamic>
  joint_matrix_mad(Group g,
    joint_matrix<Group, Ta, use::a, M, K, LayoutA> A,
    joint_matrix<Group, Tb, use::b, K, N, LayoutB> B,

>     joint_matrix<Group, Tc, use::accumulator, M, N, layout::dynamic> C);

Reviewer: Ping on this comment.  I think this issue still exists in the current version.

} // namespace sycl::ext::oneapi::experimental::matrix
```
The matrix multiply and add function performs the multiply operation
on the matrices `A` and `B`, accumulates the result with `C` and returns
the result.

==== Matrix Initialization: `joint_matrix_fill`
Unlike `joint_matrix_load` that assumes that all the matrices are
directly loaded from memory, `joint_matrix_fill`  makes it possible to
multiply a matrix which is not directly loaded from memory but rather
initialized directly in the register. On Intel AMX, if the
initialization constant is zero, this would map to the `_tile_zero`
intrinsic. Note that the value type `Tv` must be convertible to the

> matrix elements type `T`.

Author: > Since we decided to change `joint_matrix_load` and `joint_matrix_store` to have just one template type `T`, should we do the same here also and remove `Tv`?

In the case of fill, values can have other types that are convertible to T, like filling bfloat16 matrix with `5`. we use `static_cast` in the implementation so users do not need to cast.

> matrix elements type `T`.

Reviewer: OK, that makes sense, and it is consistent with `std::fill`.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, typename T, size_t Rows, size_t Cols,
          use Use, layout Layout, typename Tv>
  void joint_matrix_fill(Group g, joint_matrix<Group, T, Use,
                         Rows, Cols, Layout> &m, Tv v);

} // namespace sycl::ext::oneapi::experimental::matrix
```

> ==== Element-Wise Operations

Author: what do you mean element-wise operations is a "slang" phrase? this is the term used in BLAS library specifications

> ==== Element-Wise Operations

Reviewer: It seems like slang English to me, but I don't feel that strongly especially if it's used in other literature.  Let's be consistent, though.  The sycl_ext_intel_matrix spec use "per-element", so either change both to use "element-wise" or both to use "per-element".

> ==== Element-Wise Operations

Author: I have "per-element" in the sycl_ext_intel_matrix spec because the variant with the coordinates let you do per element access and operation.
However, the plain `joint_matrix_apply` can only do element-wise operations. You cannot really access one specific element and make special operations. 
I think it is better to make this distinction. Otherwise, why we would have a new functionality in the Intel document if it does not provide more.

> ==== Element-Wise Operations

Reviewer: I think you are saying that the term "per-element" means that the API provides coordinates while the term "element-wise" means that the API does not provide coordinates.

In my mind, the two terms are equivalent, and there is no such distinction.  I think many other readers will be the same.  If you want to make this distinction, you should be more explicit like "Per-element access with coordinates".

or adding a constant value to every element of the matrix are two
examples of this usage. When the operation depends on the element
index of the matrix, an Intel-specific extension is available as part
of the * link:sycl_ext_intel_matrix.asciidoc[sycl_ext_intel_matrix]

Besides the `Group` and the `joint_matrix` arguments,
`joint_matrix_apply` takes a C++ Callable object which is invoked once
for each element of the matrix. This callable object must be invocable
with a single parameter of type `T&`. Commonly, applications pass a
lambda expression.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template<typename Group, typename T, use Use, size_t Rows, size_t Cols,
  layout Layout, typename F>
  void joint_matrix_apply(Group g, joint_matrix<Group, T, Use, Rows, Cols,
  Layout>& C, F&& func);

} // namespace sycl::ext::oneapi::experimental::matrix
```

In the following example, every element of the matrix `C` is
multiplied by `alpha`. Then, an activation function, `relu` in this
example, is applied on each of the elements of `C`.

```c++
joint_matrix_apply(sg, C, [=](T &x) {
    x *= alpha;
    relu(x);
});
```
IMPORTANT: `joint_matrix_apply` is not implemented yet.

> === Support for `tf32` Floating Point Type

Author: Sorry for missing your comments from March 21. I think I accidentally deleted  the notification email.
Your suggestion for TF32 section make sense.
I have one doubt about the fill whether we will just use T and rely on user to cast the constant values.

The following operations (load, store, multiply-and-add, fill, and
element-wise operations) are group functions as defined in section
4.17.3 of the core SYCL specification. As such, they must be
encountered in convergent control flow by the work-items in the group
that performs the group operation. The `Group` template argument must
be `sycl::sub_group`. This extension currently supports
only the sub-group scope, so the `Group` template parameter must be
`sycl::sub_group`.

==== Load
```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, typename T,
          size_t Rows, size_t Cols,
          access::address_space Space, access::decorated IsDecorated>
void joint_matrix_load(Group g,
    joint_matrix<Group, T, use::accumulator, Rows, Cols,
    layout::dynamic> &res,
    multi_ptr<T, Space, IsDecorated> src, size_t stride, layout Layout);

// Only available when Layout != layout::dynamic
template <typename Group, typename T,
          size_t Rows, size_t Cols,
          use Use, layout Layout,
          access::address_space Space, access::decorated IsDecorated>
void joint_matrix_load(Group g,
    joint_matrix<Group, T, Use, Rows, Cols, Layout> &res,

>     multi_ptr<T, Space, IsDecorated> src, size_t stride);

Reviewer 2: Yeah at some point raw pointers should be supported: There was an idea that it would eventually be possible to cast a raw pointer to a multi_ptr which would mean that we don't duplicate all these function implementation for the USM case. But I'm not sure what the status of this is.

initialization constant is zero, this would map to the `_tile_zero`
intrinsic. Note that the value type `Tv` must be convertible to the
matrix elements type `T`.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, typename T, size_t Rows, size_t Cols,
          use Use, layout Layout, typename Tv>
void joint_matrix_fill(Group g, joint_matrix<Group, T, Use,
                         Rows, Cols, Layout> &m, Tv v);

} // namespace sycl::ext::oneapi::experimental::matrix
```

==== Element-Wise Operations
Besides matrix multiply and add, this extension aims to make it
possible to perform element-wise operations on matrices in a SPMD
manner. `joint_matrix_apply` function performs an element-wise
operation where the same operation is performed on every element of
the joint matrix, such that the operation can be performed without knowledge
of the position of the element within the matrix. Activation functions
or adding a constant value to every element of the matrix are two
examples of this usage. When the operation depends on the element
index of the matrix, an Intel-specific extension is available as part

> of the * link:sycl_ext_intel_matrix.asciidoc[sycl_ext_intel_matrix]

Author: Correct, it is a typo. Thanks!
We can keep it in the oneapi extension. Remove it if this makes it to the SYCL spec.

precision. In order to declare a `joint_matrix` object with this
element type, use `matrix::precision::tf32` in place of the `T`
template parameter.

```c++
namespace sycl::ext::oneapi::experimental::matrix::precision {

class tf32;

} // namespace sycl::ext::oneapi::experimental::matrix::precision
```

For example:

```c++
joint_matrix<sub_group, precision::tf32, use::a, tM, tK,
             layout::row_major> tA;
```

Whenever the application loads, stores, fills, or accesses the
elements of a TF32 matrix, the application sees the elements as
float. There are special overloads of these functions for TF32 for
this purpose.

==== TF32 load
These overloads of `joint_matrix_load` load float values into a TF32
matrix. It is unspecified whether the implementation loads all 32 bits
into the joint matrix or if it only loads the relevant 19 bits.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template <typename Group, size_t Rows, size_t Cols,
          access::address_space Space, access::decorated IsDecorated>
void joint_matrix_load(Group g,
    joint_matrix<Group, precision::tf32, use::accumulator, Rows, Cols,
    layout::dynamic> &res,

>     multi_ptr<float, Space, IsDecorated> src, size_t stride, layout Layout);

Reviewer: ```suggestion
    multi_ptr<const float, Space, IsDecorated> src, size_t stride, layout Layout);
```

Keep this overload in sync with the main definition above.  We decided recently that the `multi_ptr` should be `const` for loads, so do that here too.  Same for the other overload below.

Joint matrix is a SYCL extension for matrix hardware programming. It
unifies targets like Intel AMX in CPUs, Intel XMX in Intel GPUs and
Nvidia Tensor Cores. This provides portable but performant API for
users who want to build their own neural networks applications,
perform custom optimzations, or experiment with new operations in a
timely and performing manner.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification. An implementation supporting this extension must predefine
the macro `SYCL_EXT_ONEAPI_MATRIX` to one of the values defined in the
table below. Applications can test for the existence of this macro to
determine if the implementation supports this feature, or applications
can test the macro's value to determine which of the extension's
features the implementation supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Matrix API Versions

While this document presents the core API that unifies Intel AMX,
Intel XMX, and Nvidia Tensor Cores, the implementations support

> slightly different versions of the API. For this reason, we introduce

Reviewer 2: The situation is different now from what this paragraph states; because this document is specifically for the unified matrix interfaces which are portable. So I think it is best to replace this paragraph completely with e.g. the standard template for feature macro versioning.

Author: @JackAKirk, the standard template for feature macro versioning if not for experimental feature.
Once this moves out of experimental and becomes supported, this whole section "Matrix API versions" and SYCL_EXT_ONEAPI_MATRIX_VERSION macro will be removed. We won't need to keep the legacy API and tests. Right now, we only keep them to ensure current users have something working while we guide them through all these changes until we have something final (hopefully in this PR).

Do you suggest I remove this now?

Reviewer 2: I just think that anyone reading this now as documentation on `joint_matrix` will be thrown by this

"the implementations support
slightly different versions of the API"

And since this does seem to be the main place that people will arrive at for `joint_matrix` documentation currently, it makes sense to address this now.

Author: I will remove it especially that we made value 4 as the default already.

> }

Author: yes, this should be just: "struct joint_matrix;"

```

IMPORTANT: Matrix layout defaulting to `layout::dynamic` applies only

> to `joint_matrix` with `use::accumulator`

Author: >>> As I read through the spec more, it seems like a matrix declared with use::accumulator must use layout::dynamic. Is that the case?
Yes that is correct; I will change the text as you suggested.

==== Use

> The main operation performed by the matrix hardware is `D=C+A*B`. `Use`

Reviewer 2: We don't actually allow D!=C in the interface. So this could be a little confusing unless we clarify this distinction between what hardware can do and what the interface exposes? Should we really say "main operation performed by the matrix extension" and "C=C+A*B" and so on?

Reviewer: This seems like an important issue to resolve.  I remember asking early on whether we could really support the case when `D != C`, and I thought the answer was "yes".  I think this was before we decided to support CUDA with this API, though, so perhaps the answer is different for CUDA.

If we cannot support this case, then I think we need to change the API to make it impossible for the application to ask for something that is not supported.  For example, maybe `joint_matrix_mad` should be defined to modify the `C` parameter in-place.  Does `joint_matrix_mad` need to return anything in this case?

Reviewer 2: Sorry I am plain wrong here. I was thinking about the fact that the types of the returned matrix D and the C must be the same joint_matrix type according to the interface. In practice this often means that the usage is that C==D but the API already allows that C!=D. I think it is fine as it is and you can ignore my thoughtless comment.

Author: I confirm that we support C!=D for all backends including CUDA.  So this should not be an issue.

}
```

IMPORTANT: Matrix layout defaulting to `layout::dynamic` applies only

> to matrix with `use::accumulator`

Reviewer 2: ```suggestion
to `joint_matrix` with `use::accumulator`
```

==== Use

> Specifying the usage of the matrix: matrix left (A), matrix right (B)

Reviewer 2: `matrix left` and `matrix right` are not defined (or equivalently A / B aren't defined).

}
```

==== Group Memory Scope
In this API, we use the terminology of `joint_matrix` instead of plain
`matrix` to emphasize that the matrix is shared among a group of work
items and is not private to each work item. The group scope is added
as an additional template parameter.

IMPORTANT: In the current implementation, only the `sub_group` scope

> is supported

Author: No I don't anticipate work group support to be done before that. This is similar to sizes and types that are implementation defined. That's why we provide scopes query in the device descriptors.

> is supported

Reviewer: In this case, I think it would be better to remove these IMPORTANT notes and simply state that the `Group` template parameter must be `sub_group`.  You can state this requirement in the section "Group Memory Scope".

As we discussed in the SPV meeting, it easy to relax requirements in a future version of the spec.  If we decide to support other groups in the future, it's easy to modify the spec to say this.

You raise an interesting point about the query. Presumably, all devices that support joint matrix support the `sub_group` execution scope right now, and they don't support any other execution scope.  Therefore, I think it makes sense to remove that query from this version of the spec.  If we add other execution scopes in future versions of the spec, we can decide at that time whether it makes sense to add a query.  For example, if all devices support the same set of execution scopes, there's no need to ever add a query.  A query is only useful if the set of execution scopes is different from some devices, and it's not clear at this point whether that will be the case.

> is supported

Author: Yes, we should be able to remove these notes and state it as you suggested since the only scope currently supported is subgroup.

@JackAKirk, can you please confirm that's the expectation from the CUDA backend as well? 

I will also remove the scope query.

When the group is a `sycl::sub_group`, a matrix is declared as follows:

```c++
joint_matrix<sub_group, int8_t, use::a, tM, tN, layout::row_major> tA;
```

=== Matrix Operations and their Execution Scope
We define three new functions needed to perform the main and common
operations on matrices, namely load, store, and the actual multiply
and add operation. This set of functions can be easily extended if the
matrix hardware implements new features.

Since the matrix functions are group operations (as defined in Section
4.17.3 of the SYCL specification), the matrix API has to be accessed
by all the work-items in the group in a convergent control flow. The
`Group` template argument can be a work-group or a sub-group. These

> functions will be called once by each work item in the group.

Author: Yes, I will just remove it.

To be aligned with the SYCL 2020 group algorithms, an additional group
argument is added to the matrix operations to designate that these
functions are collective operations. The {dpcpp} syntax is the following:

IMPORTANT: In the current implementation, only the `sub_group` scope
is supported.

==== Load
```c++

> namespace sycl::ext::oneapi::experimental::matrix {

Author: yes, I will add that

  template <typename Group, typename T, typename S,
            size_t NumRows, size_t NumCols,

>             access::address_space Space>

Author: good catch, thanks!

  void joint_matrix_load(Group sg,
    joint_matrix<Group, T, use::accumulator, NumRows, NumCols,
    layout::dynamic> &res,
    multi_ptr<S, Space, IsDecorated> src, size_t stride, layout Layout);
    
  template <typename Group, typename T, typename S,
            size_t NumRows, size_t NumCols,
            use Use, layout Layout, access::address_space Space>
  void joint_matrix_load(Group sg,
    joint_matrix<Group, T, Use, NumRows, NumCols, Layout> &res,
    multi_ptr<S, Space, IsDecorated> src, size_t stride);
}
```

`joint_matrix_load` loads data from memory to the 2d tiles/registers
of the matrix hardware.
We define two overloads of the load function depending on whether the
memory layout was declared as part of the `joint_matrix` type or not. 
The first overload that takes memory layout as an argument is only
available for a `joint_matrix` type that used the default value
`layout::dynamic`.
The second overload without a memory layout must not be used with a
`joint_matrix` type that used the default value `layout::dynamic`.

The base pointer `src` here determines the starting address of the
matrix to be loaded from. `Layout` determines whether the data is

> being read in a row (`row_major`), column major (`column_major`)

Author: That is correct.

fashion. `stride` describes the number of elements between consecutive
rows for the row major layout, or between columns for the column major

> layout.

Author: Right, this is specific to the tf32 case where the elements of memory are float but type of joint matrix is precision::tf32. SO these can be different.
In the store case, since we only supports store of accumulator matrix which is always full precision, so no separation of types there.

> layout.

Reviewer: > In the store case, since we only supports store of accumulator matrix which is always full precision

I didn't realize this.  Does the spec say someplace that `tf32` is not allowed for the accumulator matrix?

> Right, this is specific to the tf32 case where the elements of memory are float but type of joint matrix is precision::tf32.

If this is the only reason, then I don't think you need separate `T` and `S` template parameters.  You could instead use the "alias" mechanism you describe in your other comment, or you can define separate overloads for the case when `T` is `tf32`.

> layout.

Author: @JackAKirk, can you please confirm this is the only reason for the CUDA backend to have type of matrix different from type of memory elements?
if this is the case, the recommendation is to create overloads for tf32. @JackAKirk, Is that okay for the CUDA backend as well?

the joint matrix, such that the operation can be performed without knowledge
of the position of the element within the matrix. Activation functions
or adding a constant value to every element of the matrix are two
examples of this usage. When the operation depends on the element
index of the matrix, an Intel-specific extension is available as part
of the * link:sycl_ext_intel_matrix.asciidoc[sycl_ext_intel_matrix]

Besides the `Group` and the `joint_matrix` argument,
`joint_matrix_apply` takes a lambda expression as an argument that
specifies the specific operation on each of the elements of the input
matrix.

```c++
namespace sycl::ext::oneapi::experimental::matrix {
  template<typename Group, typename T, use Use, size_t M, size_t N,
  layout Layout, typename F>
  void joint_matrix_apply(Group g, joint_matrix<Group, T, Use, M, N,
  Layout>C, F&& lambda);
}
```

In the following example, every element of the matrix `C` is
multiplied by `alpha`. Then, an activation function, `relu` in this
example, is applied on each of the elements of `C`.

```c++
joint_matrix_apply(sg, C, [=](T x) {
    x *= alpha;
    relu(x);
});
```
IMPORTANT: `joint_matrix_apply` is not implemented yet.

> === Joint Matrix Additional Types

Author: > The title of this section should be something like "Support for TF32 floating point" since that is the only other 

Sounds good
 
> * It seems like the purpose of this support is only to reduce the precision of the matrix MAD operation.  As far as the application is concerned, the values appear to be `float`.  When you load values into the matrix, they are `float`.  When you store values out of the matrix, they are `float`.  I presume the per-element access APIs also return `float` values.  Is this a correct understanding?  If so, it would be good to make this clearer in the description somehow.

I say that in the last paragraph but I will clarify more.

> * I assume the application declares a TF32 joint matrix by using the `precision::tf32` type for the `T` template parameter, correct?  The text in this section doesn't actually say that, but it seems like that must be the intent.  Again, it would be good to state this more clearly.  Maybe an example would help too.

Will add that, thanks!

> * I assume there must be overloads for all the joint matrix functions for the `precision::tf32` type, which are different from the ones documented above.  For example, I presume there must be an overload of `joint_matrix_store` like this:
>   If this is the intent, all these overloads should be listed in this section.

Is this really necessary?
In the [actual implementation](https://github.com/intel/llvm/pull/8151/), I did not use overloads per say, but I used traits as follows:
template <typename T> struct helper_traits {
  using element_type = T;
  using storage_element_type = T;
  using fill_argument_type = T;
};

template <> struct helper_traits<precision::tf32> {
  using element_type = precision::tf32;
  using storage_element_type = float;
  using fill_argument_type = float;
};

So in the API, I use storage_element_type and element_type instead of overloads.
In this case, can we just say in the text that  the functions load, store, fill, mad, apply work on joint matrix of type tf32 as well?

> * When the application calls `joint_matrix_load`, I assume the implementation implicitly rounds the `float` value to TF32.  Correct?  What rounding mode is used for this operation?  Round toward zero, round up, round down, etc.?  I presume this should be specified.

That is not true: no rounding is implicitly happening.  As we say before, everything is manipulating floats, except joint_matrix_mad. If users call joint_matrix_mad with no explicit rounding, truncation will happen (only the 19 bits will be taken into account). I will add a clarification about this in the text.

> * When that application calls `joint_matrix_apply`, I presume the callback function receives a value of type `float &`?  This should be clarified.  If the callback function modifies the value, I presume the implementation implicitly rounds the `float` value to TF32 before storing it back to the matrix.  Again, clarify which rounding mode is used.

Same applies to joint_matrix_apply, input joint matrix is tf32, but actual data are floats (storage). no implicit rounding is happening.  

> * Since it seems like there are all these implicit rounding operations that occur, when would the application need to use `round_to_tf32`?  If there is no compelling use case, can we remove that function?

No implicit rounding is happening. data is float.  If users want data to be actually tf32, they have to use the rounding function.

Besides C++ `half`, `float`, `double` types, and `sycl::bfloat16` types, joint
matrix implementations may support other low-precision floating-point types
such as tf32. tf32 type has a 19 bit format with one sign bit, 8
exponent bits offering the same range as fp32,  and 10 mantissa bits
offering same precision as  half type. The usage of tf32 type is
restricted to `joint_matrix` using:
`sycl::ext::oneapi::experimental::matrix::precision::tf32`.

Joint matrix type tf32 is defined as an empty class with no member functions.
```c++
namespace precision {
  class tf32;
}
```
Besides the type, one conversion function is added:
`round_to_tf32` that  performs the rounding to tf32.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

>   float round_to_tf32(float &elem);

Author: The rounding mode is implementation defined and depends on every hardware. So far, I don't think hardware exposes this choice to users for tf32 and/or which mode is public information.
Do we provide the mode parameter for other types conversion function?

inform them about supported types, sizes, scope, and layouts by the
implementation. This also offers development and tuning productivity
by both scientists and library developers. We provide two types of the
query interface: compile-time query and runtime query.

==== Compile-Time Query
This returns `constexpr` values to use in `joint_matrix` template
arguments but depends on an enumeration of the matrix hardware that
can be tested.  The compile-time query interface proposed here
consists of two functionalities:

- Validation: at compile time, the validation functionality informs
  the user whether a specific combination is valid or not. This takes
  place when the user specifies all template parameters.

- Default values: this provides a default shape if the user does not
  provide a specific combination. In this case, aliases to the
  `joint_matrix` type can be used, namely
  `joint_matrix_a/b/accumulator` where no additional argument is
  needed. This form happens when the user specifies all template
  parameters except the sizes of the matrices (`tiles`) M, N, and K.

The table below provides a description for each of the member

> variables in `matrix_params` class and the forms in which  they are

Author: I had the same preference.

defined.

[frame="none",options="header"]
|======================
| Member/type alias in `tpu_params` | Description
|`type_a`| type alias for the type of matrix A

> |`type_b`| type alias for the type of matrix B

Author: these are type of the matrix itself so it should be tf32 to match joint_matrix type which is precision::tf32

|`type_accumulator`| type alias for the type of matrix accumulator
|`M`|when no sizes are provided by the user, indicates the suggested
default size for M; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value M that the user provides if M is
supported by the implementation
|`N`|when no sizes are provided by the user, indicates the suggested
default size for N; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value N that the user provides if N is
supported by the implementation
|`K`| when no sizes are provided by the user, indicates the suggested
default size for K; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value K that the user provides if K is
supported by the implementation
|`joint_matrix_a`| type alias for `joint_matrix` for matrix A
|`joint_matrix_b`| type alias for `joint_matrix` for matrix B |
`joint_matrix_accumulator`| type alias for `joint_matrix` for matrix

> accumulator

Reviewer: These three type aliases are templates, so you should show them that way.

Author: Do you mean I show the full definition as follows:
`template <typename Group, layout Layout> using joint_matrix_a;`
?

Reviewer: Yes

|======================

```c++
namespace sycl::ext::oneapi::experimental::matrix {

template<tpu u, typename Ta=void, typename Tb=void, typename Tc=void,

> int sM=0, int sN=0, int sK=0>

Author: Good catch

  static constexpr size_t M = /* implementation defined */;
  static constexpr size_t N = /* implementation defined */;
  static constexpr size_t K = /* implementation defined */;

  template <typename Group, layout Layout>
  using joint_matrix_a = joint_matrix<Group, Ta, use::a, M, K, Layout>;

  template <typename Group, layout Layout>
  using joint_matrix_b = joint_matrix<Group, Tb, use::b, K, N, Layout>;

  template <typename Group>
  using joint_matrix_accumulator = joint_matrix<Group, Taccumulator,
  use::accumulator, M, N>;
};

} // namespace sycl::ext::oneapi::experimental::matrix
```
===== Validation Example:
```c++
// User can provide sizes besides the types and matrix_params can assert
// if they are supported or not
// in this case, an assertion will happens as 16 is not a supported size for M
using myparams =
matrix_params<sycl::ext::oneapi::experimental::architecture::intel_gpu_pvc,
int8_t, int8_t, int, 16, 16, 32>;
size_t NDRangeM = M / myparams::M;  //Assertion would happen at this line
size_t NDRangeN = N / myparams::N;
```

===== Default Values Example:
```c++
using myparams =
matrix_params<sycl::ext::oneapi::experimental::architecture::intel_gpu_pvc,

> int8_t, int8_t, int>;

Author: We can keep default for `Td` but I don't have a preference.

default size for M; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value M that the user provides if M is
supported by the implementation
|`N`|when no sizes are provided by the user, indicates the suggested
default size for N; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value N that the user provides if N is
supported by the implementation
|`K`| when no sizes are provided by the user, indicates the suggested
default size for K; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value K that the user provides if K is
supported by the implementation
|`template <typename Group, layout Layout> using joint_matrix_a;`| type
alias for `joint_matrix` for matrix A
|`template <typename Group, layout Layout> using joint_matrix_b;`| type
alias for `joint_matrix` for matrix B
|`template <typename Group> using joint_matrix_accumulator;`| type

> alias for `joint_matrix` for matrix accumulator

Reviewer: Please adjust the column width in this table to avoid line wrap.

> alias for `joint_matrix` for matrix accumulator

Reviewer: The column widths were not changed in your recent commit bf8e00c.  The column widths are defined by the "col" option in the table definition.  For example, change the line:

> [frame="none",options="header"]

to something like:

> [frame="none",options="header",cols="30%,70%"]

The percentages indicate the fraction of each column's width in the table (the numbers must add to 100).  You'll have to play around with the numbers to avoid the line wrap for the `template<typename Group>` or `template<typename Group, layout Layout>`.  To check your results, look at the HTML render in GitHub.

> alias for `joint_matrix` for matrix accumulator

Reviewer: Everything looks good, but this one comment has not been addressed yet.

> alias for `joint_matrix` for matrix accumulator

Reviewer: Fixed in ee282500c36070319828ff5d62e585f5fa5e887d

|======================

```c++
namespace sycl::ext::oneapi::experimental::matrix {

// This is the validation form, when all template parameters are
// specified.
template<sycl::ext::oneapi::experimental::architecture Dev, typename
Ta=void, typename Tb=void, typename Tc=void, typename Td=Tc, size_t sM=0,

> size_t sN=0, size_t sK=0>

Author: This is correct; I am making the change.

size_t NDRangeN = N / myparams::N;
//if M, N, K do not multiply the default sizes, padding has to be done
// device code: the matrices are constructed using the default dimensions
myparams::joint_matrix_a<sub_group, layout::row_major> sub_a;
myparams::joint_matrix_b<sub_group, layout::row_major> sub_b;
myparams::joint_matrix_c<sub_group> sub_c;

```
==== Runtime Query
This provides a more general query interface with information about
sizes and types that are supported by a specific matrix
implementation. This is needed to avoid padding by the user, for
tuning, and efficient code generation if used by a library.

The table below provides a description for each of the device matrix
descriptors that can be queried using `get_info` API.

[frame="none",options="header"]
|======================
| Device descriptors | Return type| Description
|`ext::oneapi::experimental::info::device::matrix::combinations` |
`std::vector<combination>`| tells the set of supported matrix sizes
and types on this device
|======================

The general query returns a vector of `combinations` of `combination`
type. Each combination includes the sizes and the types for the

> matrices A, B, and accumulator. Note that for each matrix hardware,

Reviewer: ```suggestion
matrices A, B, C, and D. Note that for each matrix hardware,
```

the query returns `max_msize, max_nsize, max_ksize` or `msize, nsize,
ksize` exclusively, depending on whether the implementation supports a
continuous or discrete number of sizes. If a device support a
continuous number of sizes, the `max_*` variant is applied and only
the maximum number is returned. However, if a device supports a
discrete list of numbers so the `msize, nsize, ksize` variant is applied.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

enum class matrix_type {
  bf16,
  fp16,
  tf32,
  fp32,
  fp64,
  sint8,
  sint16,
  sint32,
  sint64,
  uint8,
  uint16,
  uint32,
  uint64
};
struct combination {
  uint32_t max_msize;
  uint32_t max_nsize;
  uint32_t max_ksize;
  uint32_t msize;
  uint32_t nsize;
  uint32_t ksize;
  matrix_type atype;
  matrix_type btype;
  matrix_type ctype;
  matrix_type dtype;
};

} // namespace sycl::ext::oneapi::experimental::matrix
```

Each combination of the `combinations` vector composes the types and

> sizes of A, B, accumulator matrices supported by the device

Reviewer: ```suggestion
sizes of A, B, C, and D matrices supported by the device
```

implementation. The
table below provides a description of each member of the `combination` struct.

[frame="none",options="header"]
|======================
| Member of `combination` | Description
|`max_msize`, `max_nsize`, `max_ksize`| if the matrix implementation
supports a continuous number of element sizes, each of these members
is non-zero, and the matrix implementation supports all element sizes
from 1 up to (and including) that number. By contrast, if the matrix
hardware implementation supports a discrete number of element sizes,
each of these members has the value zero
|`msize`, `nsize`, `ksize`| if the matrix implementation supports a
discrete number of element sizes, each of these members is non-zero,
and the value tells one of the supported element sizes. By contrast,
if the matrix hardware supports a continuous number of element sizes,
each of these members has the value zero
|`atype`, `btype`, `ctype`, `dtype`| indicates the types supported in
the combination. these are of type `matrix_type` which tells the list

> of types that are supported for the A, B, and accumulator matrices in

Reviewer: ```suggestion
of types that are supported for the A, B, C, and D matrices in
```

`sint64`: `int64_t` +
`uint8`: `uint8_t` +
`uint16`: `uint16_t` +
`uint32`: `uint32_t` +
`uint64`: `uint64_t`
|======================

===== Runtime Query Example:
```c++
// Ta, Tb, Tc, and Td are the types used in applications
std::vector<combination> combinations =
           device.get_info<info::device::matrix::combinations>();
for (int i = 0; sizeof(combinations); i++) {
  if (Ta == combinations[i].atype &&
      Tb == combinations[i].btype &&
      Tc == combinations[i].ctype &&
      Td == combinations[i].dtype) {
    // joint matrix GEMM kernel can be called using these sizes
    joint_matrix_gemm(combinations[i].msize,
         combinations[i].nsize, combinations[i].ksize);
  }
}
```

=== Appendix: Supported Combinations Per Hardware

The table below provides a list of the combinations that
`joint_matrix` implementations support on each of Intel AMX and Intel
XMX hardware. Note that these can be returned using
`ext::oneapi::experimental::info::device::matrix::combinations`.

==== Intel AMX Supported Combinations
This is currently available in
`sycl::ext::oneapi::experimental::architecture::intel_cpu_spr`.

[frame="none",options="header"]
|======================

> | A type | B type | Accumulator type (`ctype` = `dtype`) | M | N | K

Reviewer: The name of the column "Accumulator type (`ctype` = `dtype`)" seems confusing to me.  I'd suggest the following:

* Change the column name to "C and D type".
* Change the introduction paragraph to say:

  > This is currently available in devices with the architecture `architecture::intel_cpu_spr`.  In this architecture, the C and D matrices support the same set of types, so they are shown in a single column.

Make a similar change below in the table for the section "Intel XMX Supported Combinations".

Reviewer: Thinking about this more ... I suppose the type of C and the type of D must be the *same*, right?  If so, the intro should be more like this:

> This is currently available in devices with the architecture `architecture::intel_cpu_spr`. In this architecture, the type of the C matrix must be the same as the type of the D matrix.  Therefore, that common type is shown in a single column in the table below.

(I'm not sure how you want to add the word "implementation", but you can add it in this paragraph too if you want.)

default size for N; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value N that the user provides if N is
supported by the implementation
|`K`| when no sizes are provided by the user, indicates the suggested
default size for K; usually this corresponds to the maximum size the
implementation supports. In validation mode, where the user does
provide sizes, this is the same value K that the user provides if K is
supported by the implementation
|`template <typename Group, layout Layout> +
using joint_matrix_a;`| type alias for `joint_matrix` for matrix A
|`template <typename Group, layout Layout> +
using joint_matrix_b;`| type alias for `joint_matrix` for matrix B
|`template <typename Group> +
using joint_matrix_c;`| type alias for `joint_matrix` for the input matrix accumulator
|`template <typename Group> +
using joint_matrix_d;`| type alias for `joint_matrix` for the output matrix accumulator
|======================

```c++
namespace sycl::ext::oneapi::experimental::matrix {

// This is the validation form, when all template parameters are
// specified.

> template<architecture Arch, typename Ta, typename Tb, typename Tc,
> typename Td, size_t sM, size_t sN, size_t sK>

Reviewer: ```suggestion
template<architecture Arch, typename Ta, typename Tb, typename Tc,
typename Td, size_t sM=0, size_t sN=0, size_t sK=0>
```

You need these default values here if you want the application to be able to use the "default values form" without explicitly providing `0`'s for these template parameters.

Author: if sM,sN,sK values are zero, M,N,K return also zero, and it will generate an assertion in this case as zeros is not supported by any hardware. So in the validation form, sM,sN,sK are required.
Are you saying we cannot use the "default values form" specialization of the query if we don't specify zeros here?

Author: I think you are talking about the primary definition. I will add defaults there.
This one is a specialization and should not have the defaults.

Reviewer: I think the "validation form" is the primary template definition.  In any case ... yes, the default values need to be specified on the primary template definition (not the specialization).

struct matrix_params {
  // An implementation typically uses static_assert here to trigger a
  // compilation error when the matrix types or shapes are not
  // supported by the device identified by the architecture "Arch".

  static constexpr size_t M = sM;
  static constexpr size_t N = sN;
  static constexpr size_t K = sK;

  template <typename Group, layout Layout>
  using joint_matrix_a = joint_matrix<Group, Ta, use::a, sM, sK, Layout>;

  template <typename Group, layout Layout>
  using joint_matrix_b = joint_matrix<Group, Tb, use::b, sK, sN, Layout>;

  template <typename Group>
  using joint_matrix_c = joint_matrix<Group, Tc, use::accumulator, sM, sN>;

  template <typename Group>
  using joint_matrix_d = joint_matrix<Group, Td, use::accumulator, sM, sN>;
};

// This is the default values form, where the matrix dimensions are
// omitted.

> template<architecture Arch, typename Ta, typename Tb, typename Tc, typename Td=Tc>

Reviewer: ```suggestion
template<architecture Arch, typename Ta, typename Tb, typename Tc, typename Td>
```

C++ does not allow you to provide a default template parameter for a template specialization, so the `=Tc` part will cause a compilation failure.  If you want `Td` to have a default value, it must be specified in the primary template (i.e. the first one defined in the synopsis).

Author: I think you are looking at the wrong synopsis. 
The current version is missing the primary template so I am adding that. I will specify the default there.

      (sM == 0 && sN == 0 && sK == 0) ||
          (is_combination_valid_amx<Ta, Tb, Tc>(sM, sN, sK)),
      "Invalid parameters for Intel AMX, query valid types and maximum sizes "
      "using: dev.get_info<experimental::info::device::matrix>(); and
      then check out matrix::combinations array");
  using type_a = Ta; // this type alias is not available in the
  current implementation 
  using type_b = Tb; // this type alias is not available in the
  current implementation
  using type_accumulator = Tc; // this type alias is not available in

>   the current implementation

Author: I will remove these comments and address them as part of the changes I should make to the static query in general (like change name from `tpu_params` to `matrix_params`)

  using joint_matrix_b = joint_matrix<Group, Tb, use::b, K, N, LayoutB>;
  template <typename Group>
  using joint_matrix_accumulator = joint_matrix<Group, Tc,
  use::accumulator, M, N>;
};
enum class tpu {
  xmx8,
  xmx16,
  amx
};
```
===== Validation Example:
```c++
// User can provide sizes besides the types and tpu_params can assert
  if they are supported or not
// in this case, an assertion will happens as 16 is not a supported size for M
using myparams = tpu_params<tpu::xmx16, int8_t, int8_t, int, 16, 16, 32>;
size_t NDRangeM = M / myparams::M;  //Assertion would happen at this line
size_t NDRangeN = N / myparams::N;
```

===== Default Values Example:
```c++
using myparams = tpu_params_both<tpu::xmx16, int8_t, int8_t, int>;
// use this to construct the ranges on the host side
size_t NDRangeM = M / myparams::M;
size_t NDRangeN = N / myparams::N;
//if M, N, K do not multiply the default sizes, padding has to be done
// device code: the matrices are constructed using the default dimensions
myparams::joint_matrix_a<sub_group, layout::row_major> sub_a;
myparams::joint_matrix_b<sub_group, layout::row_major> sub_b;
myparams::joint_matrix_accumulator<sub_group> sub_c;

```

> ==== Runtime Query

Reviewer: I recommend using the following outline for this section:

* Start with some introductory text about the queries.
* Table of the new info descriptors.
* Text describing the format of the vector of `combination` structs.
* Code synopsis showing `scope`, `matrix_type`, and `combination`.
* Table describing each member of `combination`.
* Example

This provides a more general query interface with information about
sizes, types,  and scopes that are supported by a specific matrix
implementation. This is needed to avoid padding by the user, for
tuning, and efficient code generation if used by a library. The
general query returns an array of `combinations` of `combination`
type. Each combination includes the sizes and the types for the
matrices A, B, and accumulator. Note that for each matrix hardware,
the query returns `max_msize, max_nsize, max_ksize` or `msize, nsize,
ksize` exclusively, depending on whether the implementation supports a
continuous or discrete number of sizes. For example, the Intel AMX
implementation supports a continuous number of sizes, so the `max_*`
variant is applied and only the maximum number is returned. The Intel
XMX implementation, on the other hand, supports a discrete list of

> numbers so the `msize, nsize, ksize` variant is applied.

Reviewer: Reword this so that it does not talk about specific Intel hardware.  Instead, talk in general terms about some devices supporting discrete sizes while others support a range.

The table below provides a description for each of the device matrix
desciptors that can be queried using `get_info` API.

[frame="none",options="header"]
|======================
| Device descriptors | Return type| Description
|`ext::oneapi::experimental::info::device::matrix::numtiles`| `uint32_t`

> |indicates number of tiles in Intel AMX (does not apply to Intel XMX)

Author: Okay, I will add it to "sycl_ext_intel_matrix" extension as this information is important for performance (if you use more than 8 tiles, you end up with spill code)

|`ext::oneapi::experimental::info::device::matrix::scopes`
|`std::vector<scope_t>`
|indicates the memory and execution scopes supported by the matrix

> implementation

Reviewer: I think this is just the execution scope, not a memory scope.  For example, there is a section above titled "Matrix Operations and their Execution Scope".  We should also reword all these descriptions to indicate they are properties of the device, since they are all device queries.  For example:

> Tells the execution scopes that are supported by the joint matrix operations on this device.

|`ext::oneapi::experimental::info::device::matrix::num_scopes`|`uint32_t`

> |indicates number of scopes supported by the matrix implementation

Reviewer: There is no need for this query because the user can get the number of scopes by asking for the size of the vector returned from `ext::oneapi::experimental::info::device::matrix::scopes`.  Same for `num_combinations`.

|`ext::oneapi::experimental::info::device::matrix::combinations` |
`std::vector<combination>`| tells the set of supported matrix sizes
and types
|`ext::oneapi::experimental::info::device::matrix::num_combinations`|`uint32_t`
|indicates number of combinations supported by the matrix
implementation which corresponds to the size of the `combinations` array

> |`combination` 

Reviewer: This is confusing because it mixes the info descriptors together in the same table as the descriptions of the struct members.  Create a new table for the struct members with just two columns "member variable" and "description".  The type is documented in the synopsis, so it doesn't need to be repeated in the table.

|`combination`
|composes the types and sizes of A, B, accumulator matrices allowed in
one combination
|`max_msize`, `max_nsize`, `max_ksize`| `uint32_t`| if the matrix
implementation supports a continuous number of element sizes, each of
these members is non-zero, and the matrix implementation supports all
element sizes from 1 up to (and including) that number. By contrast,
if the TPU implementation supports a discrete number of element
sizes, each of these members has the value zero
|`msize`, `nsize`, `ksize`| `uint32_t`| if the matrix implementation
supports a discrete number of element sizes, each of these members is
non-zero, and the value tells one of the supported element sizes. By
contrast, if the matrix hardware supports a continuous number of
element sizes, each of these members has the value zero
|`atype`, `btype`, `accumulatortype`| `matrix_type` | indicates the

> types supported in the combination

Reviewer: This would be a good place to describe what each of the `matrix_type` enumerators mean.  For example:

> Tells the combination of types that are supported for the _A_, _B_, and _accumulator_ matrices in the `T` template parameter as follows:
>
> * `bf16`: `sycl::bfloat16`
> * `fp16`: `sycl::half`
> * `tf32`: `sycl::ext::oneapi::experimental::matrix::precision::tf32`
> * `fp32`: `float`
> ...

|======================

> ```c++

Reviewer: Add a `namespace` declaration here, so users know what namespace these are defined in.

> enum class scope_t {

Reviewer: Why is this named `scope_t` with a "_t" suffix?  None of the other types have this suffix.  Can we just name it `scope` or maybe `execution_scope`?

```c++
auto data = get_wi_data(sg, C);
for (int i = 0; i < data.length(); ++i) {
  sum_of_local_rows[row] += data[i];
}
```

We want to keep backward compatibility in the joint matrix code when implementations or hardware change. To that end, instead of hard-coding this mapping, we use general backend and target-agnostic functionality, especially in the JIT compilation mode of SYCL. For this reason we would like to be able to query this mapping so that code does not have to change from one version to the other.

So for the mapping problem, since this mapping is implementation-defined, one of the proposals is to add runtime functions like:
```c++
auto data = get_wi_data(sg, C);
for (int i = 0; i < data.length; ++i) {
  auto row, col = data[i].get_coord();
  sum_of_local_rows[row] += data[i];
}
```

## TODO List

> - Add WI data to joint matrix mapping coordinates information for piece-wise operations. This will be added as part of the query or new methods to the 'get_wi_data' class. 

Reviewer 2: This feature is not possible with the "wmma.mma" ptx instructions that we use in `joint_matrix_mad` and this was a conscious decision for which we investigated the pros and cons. 

I don't think it really makes sense to have this get_coord feature in DPC++ for ext_oneapi_cuda backend at all even though it would be technically possible to force it: the only way we could do this would be to support the arch specific "mma" ptx variants of the mad operations which are tuned for a single Nvidia architecture (sm version) and each one (targetting a given sm version) uses different supported sizes. For this reason the mma variants are only currently used in high level libraries like Cutlass, and it would make sense to use them in libraries like SYCL-BLAS in the same way via inline PTX. They are not exposed in the CUDA runtime API, and I do not think it makes sense to expose them in DPC++ either: I think the "wmma.mma" option that we went for is the correct option here to maximize portability across sm versions and allow a more user friendly API.

> - Add WI data to joint matrix mapping coordinates information for piece-wise operations. This will be added as part of the query or new methods to the 'get_wi_data' class. 

Author: @gmlueck @JackAKirk this feature is important to make get_wi_data usable in a general way. Mapping is important and this is a feedback we got from our ML users and library developers. 
While I put it in the intel document to kick off the discussion, this should be in the unified document once we agree on arguments and all.
On Nvidia, this is implementable by copy to SLM.

> - Add WI data to joint matrix mapping coordinates information for piece-wise operations. This will be added as part of the query or new methods to the 'get_wi_data' class. 

Reviewer 2: Yeah, currently you can deal with it by calling `joint_matrix_store` to copy to SLM. If `get_coord` end up being in `oneapi` namespace, then I guess for portability we'd call `joint_matrix_store` in the nvptx impl of `get_coord`. Might be tricky to do this efficiently I guess, but seems hypothetical atm.

> - Add a more realistic and complete example that shows the value of the general query. 

Author: @gmlueck, @JackAKirk,  I added an example that shows the general query API in https://github.com/intel/llvm-test-suite/pull/1492, In SYCL/Matrix/joint_matrix_query_general_tpu_sel.cpp.
This example employs: combinations, `max_msize`, `msize`, `atype`, `btype`, `accumulatortype`, and `scopes`.
 All these are needed to query a specific combination on both AMX and XMX.
I don't see a field that can be left out for both functionality and performance.
The only field I don't show in the other PR example is `num_tiles` for AMX. This is also important and was used in the AMX performance kernels. The idea is illustrated here. In order to decide how  many joint matrices (AMX tiles) to use in the innermost loop body, the user needs to know that the maximum is 8 on AMX. Otherwise you end up with spills and bad performance.  In these performance kernels, one option is to use the following code that uses total of 8 tiles but with tile reuse.  In pseudo code:
// 4 C tiles
joint_matrix<> C00;
joint_matrix<> C01;
joint_matrix<> C10;
joint_matrix<> C11;
for k loop
// 2 A tiles
  joint_matrix<> A0;
  joint_matrix<> A1;
// 2 B tiles
  joint_matrix<> B0;
  joint_matrix<> B1;
//loads of two A and two B
  C00 = joint_matrix_mad(C00, A0, B0);
  C01 = joint_matrix_mad(C01, A0, B1);
  C10 = joint_matrix_mad(C10, A1, B0);
  C11 = joint_matrix_mad(C11, A1, B1);
} //end k loop
//stores of the 4 C matrices

  template <typename Group, layout LayoutB>
  using joint_matrix_b = joint_matrix<Group, Tb, use::b, K, N, LayoutB>;
  template <typename Group>
  using joint_matrix_accumulator = joint_matrix<Group, Tc,
  use::accumulator, M, N>;

> };

Reviewer: The synopses above have too many implementation details.  I suggest defining it something like:

```
// This is the validation form, when all template parameters are specified.
template<tpu Dev, typename Ta, typename Tb, typename Taccumulator, int sM=0, int sN=0, int sK=0>
struct matrix_params {
  // An implementation typically uses static_assert here to trigger a compilation error when
  // the matrix types or shapes are not supported by the device identified by "Dev".

  using type_a = /* implementation defined */;
  using type_b = /* implementation defined */;
  using type_accumulator = /* implementation defined */;

  static constexpr size_t M = sM;
  static constexpr size_t N = sN;
  static constexpr size_t K = sK;

  template <typename Group, layout Layout>
  using joint_matrix_a = joint_matrix<Group, Ta, use::a, sM, sK, Layout>;

  template <typename Group, layout Layout>
  using joint_matrix_b = joint_matrix<Group, Tb, use::b, sK, sN, Layout>;

  template <typename Group>
  using joint_matrix_accumulator = joint_matrix<Group, Taccumulator, use::accumulator, sM, sN>;
};

// This is the default values form, where the matrix dimensions are omitted.
template<tpu Dev, typename Ta, typename Tb, typename Taccumulator>
struct matrix_params<Dev, Ta, Tb, Taccumulator, 0, 0, 0> {
  // An implementation typically uses static_assert here to trigger a compilation error when
  // the matrix types are not supported by the device identified by "Dev".

  using type_a = /* implementation defined */;
  using type_b = /* implementation defined */;
  using type_accumulator = /* implementation defined */;

  static constexpr size_t M = /* implementation defined */;
  static constexpr size_t N = /* implementation defined */;
  static constexpr size_t K = /* implementation defined */;

  template <typename Group, layout Layout>
  using joint_matrix_a = /* implementation defined */;

  template <typename Group, layout Layout>
  using joint_matrix_b = /* implementation defined */;

  template <typename Group>
  using joint_matrix_accumulator = /* implementation defined */;
};
```

Actually, I'm not sure about this part:

```
  using type_a = /* implementation defined */;
  using type_b = /* implementation defined */;
  using type_accumulator = /* implementation defined */;
```

It depends on how you answer my question above about the meaning of these type aliases.

Author: Since:
using type_a = Ta;
using type_b = Tb;
using type_accumulator = Taccumulator;

I am wondering why we needed to add these alias in the first place.

Reviewer: OK, let's remove then for now.  We can add them back later if they seem useful.

enum class tpu {
  xmx8,
  xmx16,
  amx

> };

Reviewer: Can we remove this and use `sycl::ext::oneapi::experimental::architecture` instead?
See [sycl_ext_oneapi_device_architecture](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc).

Author: [sycl_ext_oneapi_device_architecture](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc). does not have an enum for the different CPU generations.

Also, the enum we have now for the matrix hardware does not necessarily grow with number of generations of devices unless there is a drastic change in the matrix hardware. For instance, for DG2 and ATS-M, xmx8 will be used for both.

Lastly, the main problem seems to be with creating this dependency on an another experimental extension.  sycl_ext_oneapi_device_architecture states that it only works for AOT compilation and there are limitations with the implementation. It seems to be risky to depend on that.

Reviewer: > Lastly, the main problem seems to be with creating this dependency on an another experimental extension. sycl_ext_oneapi_device_architecture states that it only works for AOT compilation and there are limitations with the implementation. It seems to be risky to depend on that.

This should not be a concern.  The enum has no limitations about AOT vs. JIT.  It's just an enum.

> Also, the enum we have now for the matrix hardware does not necessarily grow with number of generations of devices unless there is a drastic change in the matrix hardware. For instance, for DG2 and ATS-M, xmx8 will be used for both.

My concern is that users won't know what generation of the matrix hardware is incorporated into each device.  User's are much more likely to know the name of the device they have.  Rather than forcing our users to know the translation of device names to matrix hardware, I think our extension should do this translation.  Our implementation can have an internal table that does this translation.

We also expect to have a query that allows the user to get the name of a device, returning one of these enums.  We don't have any similar query, which tells the user the matrix hardware generation.

I agree that some extension is needed for the CPU generations.  My plan was to add enumerations to `sycl::ext::oneapi::experimental::architecture` that follow the compilation options supported by the `opencl-aot` tool.  Last I looked these were "avx512", "avx2", "avx", and "sse4.2".  Are we adding some similar options for the matrix hardware?

Author: The compilation options when using AMX intrinsics are: -mamx-tile -mamx-int8 -mamx-bf16

Author: @gmlueck, we use -march=sapphirerapids for AOT compilation. So CPU enumeration would follow the same logic as the GPU one.

//Remainder handling
```

## Appendix: Supported Combinations Per Hardware

The table below provides a list of the combinations that `joint_matrix` implementations support on each of Intel AMX and Intel XMX hardware. Note that these can be returned in a parametrized way using the `tpu_params` query class.

### Intel AMX Supported Combinations

[frame="none",options="header"]
|======================
| A type | B type | Accumulator type | M | N | K

> | (u)int8_t  | (u)int8_t |  int32_t  |  +<=+ 16 |  +<=+ 16 |  +<=+ 64

Reviewer: Are the "type" columns supposed to be C++ type names or enumeration values from `enum class matrix_type`?  I assume they are supposed to be the enumeration values, but the names here don't match.  For example, it should be `int8` not `int8_t`.  It might be clearer to say `matrix_type::int8`.

Author: @gmlueck, i added `matrix_type::` to the types but also changed some of the types names in `matrix_type` so they match better the C++ ones. For instance int8 should be `int8_t` to not be confused with `int8` which is a vector of 8 elements of int

Author: done

myparams::joint_matrix_accumulator<sub_group> sub_c;

```
==== Runtime Query
This provides a more general query interface with information about
sizes and types that are supported by a specific matrix
implementation. This is needed to avoid padding by the user, for

> tuning, and efficient code generation if used by a library.

Reviewer: I'd recommend changing this introduction a bit.  This query isn't "more general".  The main difference is that the information isn't `constexpr` and the application doesn't need to know the type of device on which it will run.  The last sentence about "padding", etc. doesn't seem specific to this query API.  I'd say something like:

> The runtime query does not require the application to hard-code a specific device type, but it also returns values that are not `constexpr`.  It provides similar information as the compile time query API via an extended device information descriptor.

|======================
| Device descriptors | Return type| Description
|`ext::oneapi::experimental::info::device::matrix::combinations` |
`std::vector<combination>`| tells the set of supported matrix sizes
and types on this device
|======================

The general query returns a vector of `combinations` of `combination`
type. Each combination includes the sizes and the types for the
matrices A, B, and accumulator. Note that for each matrix hardware,
the query returns `max_msize, max_nsize, max_ksize` or `msize, nsize,
ksize` exclusively, depending on whether the implementation supports a
continuous or discrete number of sizes. If a device support a
continuous number of sizes, the `max_*` variant is applied and only
the maximum number is returned. However, if a device supports a
discrete list of numbers so the `msize, nsize, ksize` variant is applied.

```c++
namespace sycl::ext::oneapi::experimental::matrix {

enum class matrix_type {
  bf16,
  fp16,
  tf32,
  fp32,
  fp64,
  sint8,
  sint16,
  sint32,
  sint64,
  uint8,
  uint16,
  uint32,
  uint64
};
struct combination {
  uint32_t max_msize;
  uint32_t max_nsize;
  uint32_t max_ksize;
  uint32_t msize;
  uint32_t nsize;

>   uint32_t ksize;

Reviewer: These should all be `size_t` since the template parameters for `Rows` and `Cols` are `size_t`.

Reviewer: Ping on this comment.

is non-zero, and the matrix implementation supports all element sizes
from 1 up to (and including) that number. By contrast, if the matrix
hardware implementation supports a discrete number of element sizes,
each of these members has the value zero
|`msize`, `nsize`, `ksize`| if the matrix implementation supports a
discrete number of element sizes, each of these members is non-zero,
and the value tells one of the supported element sizes. By contrast,
if the matrix hardware supports a continuous number of element sizes,
each of these members has the value zero
|`atype`, `btype`, `accumulatortype`| indicates the types supported in
the combination. these are of type `matrix_type` which tells the list
of types that are supported for the A, B, and accumulator matrices in
the `T` template parameter as follows: +
`bf16`: `sycl::bfloat16` +
`fp16`: `sycl::half` +
`tf32`: `sycl::ext::oneapi::experimental::matrix::precision::tf32` +
`fp32`: `float` +
`fp64`: `double` +
`sint8`: signed 8 bits signed integer +
`sint16`: `signed short` +
`sint32`: `signed int` +
`sint64`: `signed long` +
`uint8`: unsigned 8 bits integer +
`uint16`: `unsigned short` +
`uint32`: `unsigned int` +

> `uint64`: `unsigned long` 

Reviewer: Strictly speaking, `short` is not necessarily a 16-bit type, `int` is not necessarily a 32-bit type, and `long` is not necessarily a 64-bit type.  It's probably better to show the fixed-width types in this list instead.  For example `sint8` corresponds to `int8_t`, and `uint64` corresponds to `uint64_t`.

|======================

===== Runtime Query Example:
```c++
// Ta, Tb, Taccumulator are the types used in applications
std::vector<combination> combinations =
           device.get_info<info::device::matrix::combinations>();
for (int i = 0; sizeof(combinations); i++) {
  if (Ta == combinations[i].atype &&
      Tb == combinations[i].btype &&
      Tc == combinations[i].accumulatortype) {
    // joint matrix GEMM kernel can be called using these sizes
    joint_matrix_gemm(combinations[i].msize,
         combinations[i].nsize, combinations[i].ksize);
  }
}
```

=== Appendix: Supported Combinations Per Hardware

The table below provides a list of the combinations that
`joint_matrix` implementations support on each of Intel AMX and Intel
XMX hardware. Note that these can be returned in a parametrized way

> using the `matrix_params` query class.

Reviewer: This is no longer true.  You can get this information through the runtime query API now.

==== Intel AMX Supported Combinations
This is currently available in
`sycl::ext::oneapi::experimental::architecture::intel_cpu_spr`.

[frame="none",options="header"]
|======================
| A type | B type | Accumulator type | M | N | K

> | `matrix_type::(u)int8`  | `matrix_type::(u)int8` |

Reviewer: These entries like "(u)int8" aren't clear.  Do you support all four combinations of signed / unsigned, or only two combinations (A and B both signed /  A and B both unsigned)?  To avoid confusion, write out all supported combinations as separate rows in the table.

Author: All 4 are supported; I will separate them.

`matrix_type::sint32`  |  +<=+ 16 |  +<=+ 16 |  +<=+ 64
|  `matrix_type::bf16`       |  `matrix_type::bf16`   |
`matrix_type::fp32`   |  +<=+ 16 |  +<=+ 16   |  +<=+ 32
|======================

==== Intel XMX Supported Combinations
This is currently available in
`sycl::ext::oneapi::experimental::architecture::intel_gpu_pvc` and
`sycl::ext::oneapi::experimental::architecture::intel_gpu_dg2`.

[frame="none",options="header"]
|======================
| A type | B type | Accumulator type | M | N | K | device
| `matrix_type::(u)int8`  | `matrix_type::(u)int8` |
`matrix_type::int32`  |  +<=+ 8 |  16 |  32 |

> sycl::ext::oneapi::experimental::architecture::intel_gpu_pvc

Reviewer: ```suggestion
architecture::intel_gpu_pvc
```

The table is hard to read with these long names.  Shortening it like this will make it easier to read.

>   sint4,

Reviewer: What are `sint2` and `sint4`?  I don't see any description above telling how to use these types in a joint matrix.  Same for `uint2` and `uint4`.

Author: These have not been added yet.
@JackAKirk are these types (2 bits, 4 bits) supported on the CUDA backend?

Reviewer 2: > These have not been added yet. @JackAKirk are these types (2 bits, 4 bits) supported on the CUDA backend?

No they won't be supported.

2 bit is unsupported on Nvidia hardware. 4 bit is supported via cuda wmma at the moment, but it is deprecated, and unsupported in sm90, so we didn't plan on adding it. It will remain supported by `mma` but that will only potentially be used in libraries. The only missing type from `joint_matrix` for the cuda backend that we may add in the future is the single-bit wmma.

Author: @gmlueck @JackAKirk, should I just remove them for now? Once we actually add support for them, I can put them back here.

  uint32,
  uint64
};
struct combination {
  uint32_t max_msize;
  uint32_t max_nsize;
  uint32_t max_ksize;
  uint32_t msize;
  uint32_t nsize;
  uint32_t ksize;
  matrix_type atype;
  matrix_type btype;
  matrix_type accumulatortype;
};
```

===== General Query Example:
```c++
constexpr int M = 1500; // with msize = 8 and msize = 4,
          // M can be broken up to 125 sequence of 8-sized ops and
          // remaining 500 using 125 sequence of 4-sized ops
auto combinations = device.get_info<info::device::matrix::combinations>();

constexpr int {msize, nsize, ksize} = break_dimension(combinations, M);
constexpr int msize_remainder = break_dimension_remainder(combinations, M);
// device code:
joint_matrix<sub_group, int8_t, use::a, msize, ksize, layout::row_major> sub_a;
joint_matrix<sub_group, int8_t, use::b, ksize, nsize, layout::row_major> sub_b;
joint_matrix<sub_group, int, use::accumulator, msize, nsize> sub_c;

> //Remainder handling

Reviewer: This is not a realistic example because of all the `constexpr`.  The `combinations` value is not `constexpr`, so the `msize`, `nsize`, and `ksize` can't be `constexpr` either.  Therefore, they cannot be used as template parameters.

Author: You are right, I will change that.
